compiler/rustc_resolve/src/late.rs

```diff
@@ -193,6 +193,13 @@ pub(crate) enum RibKind<'ra> {
     /// No restriction needs to be applied.
     Normal,
 
+    /// We passed through an `ast::Block`.
+    /// Behaves like `Normal`, but also partially like `Module` if the block contains items.
+    /// `Block(None)` must be always processed in the same way as `Block(Some(module))`
+    /// with empty `module`. The module can be `None` only because creation of some definitely
+    /// empty modules is skipped as an optimization.
```

Comment:
If the optimization is removed, some diagnostics in UI tests still change, but due to other reasons, like considering only a single immediate `Module` parent instead of walking the chain.